hibernate-core/src/main/java/org/hibernate/query/sqm/internal/AbstractSqmSelectionQuery.java

```diff
@@ -73,8 +95,169 @@ protected void errorOrLogForPaginationWithCollectionFetch() {
+	}
+
+	static class KeyedResult<R> {
+		final R result;
+		final List<Comparable<?>> key;
+		public KeyedResult(R result, List<Comparable<?>> key) {
+			this.result = result;
+			this.key = key;
+		}
+
+		public R getResult() {
+			return result;
+		}
+
+		public List<Comparable<?>> getKey() {
+			return key;
+		}
+	}
+
+	private SqmSelectStatement<KeyedResult<R>> keyed(List<Order<? super R>> keyDefinition, List<Comparable<?>> keyValues) {
+		@SuppressWarnings("unchecked")
+		final SqmSelectStatement<KeyedResult<R>> sqm =
+				(SqmSelectStatement<KeyedResult<R>>)
+						getSqmSelectStatement().copy( noParamCopyContext() );
+		final NodeBuilder builder = sqm.nodeBuilder();
+		final SqmQuerySpec<?> querySpec = sqm.getQuerySpec();
+		final SqmWhereClause whereClause = querySpec.getWhereClause();
+		final List<? extends JpaSelection<?>> items = querySpec.getSelectClause().getSelectionItems();
+		final List<SqmPath<?>> newItems = new ArrayList<>();
+		if ( items.size() == 1 ) {
+			final JpaSelection<?> selected = items.get(0);
+			for ( int i = 0; i < keyDefinition.size(); i++ ) {
+				// ordering by an attribute of the returned entity
+				if ( selected instanceof SqmRoot ) {
+					final Order<? super R> key = keyDefinition.get(i);
+					final SqmFrom<?,?> root = (SqmFrom<?,?>) selected;
+					if ( keyValues != null ) {
+						@SuppressWarnings("rawtypes")
+						final Comparable keyValue = keyValues.get(i);
+						whereClause.applyPredicate( keyPredicate( root, key, keyValue, builder ) );
+					}
+					newItems.add( root.get( key.getAttributeName() ) );
+				}
+				else {
+					throw new IllegalQueryOperationException("Select item was not an entity type");
+				}
+			}
+			sqm.select( builder.construct((Class) KeyedResult.class,
+					asList(selected, builder.construct(List.class, newItems)) ) );
+			return sqm;
+		}
+		else {
+			throw new IllegalQueryOperationException("Query has multiple items in the select list");
+		}
+	}
+
+	private <C extends Comparable<? super C>> SqmPredicate keyPredicate(
+			SqmFrom<?, ?> selected, Order<? super R> key, C keyValue, NodeBuilder builder) {
+		if ( !key.getEntityClass().isAssignableFrom( selected.getJavaType() ) ) {
+			throw new IllegalQueryOperationException("Select item was of wrong entity type");
+		}
+		final Expression<? extends C> path = selected.get( key.getAttributeName() );
+		// TODO: use a parameter here and create a binding for it
+//		@SuppressWarnings("unchecked")
+//		final Class<C> valueClass = (Class<C>) keyValue.getClass();
+//		final JpaParameterExpression<C> parameter = builder.parameter(valueClass);
+//		setParameter( parameter, keyValue );
+		switch ( key.getDirection() ) {
+			case ASCENDING:
+				return builder.greaterThan( path, keyValue );
+			case DESCENDING:
+				return builder.lessThan( path, keyValue );
+			default:
+				throw new AssertionFailure("Unrecognized key direction");
+		}
+	}
+
+	@Override
+	public KeyedResultList<R> getKeyedResultList(KeyedPage<R> keyedPage) {
```

Comment:
Something that is IMO still missing but highly desirable is the ability to specify the "scroll direction". Think of keyset pagination as scrolling in terms of getting results that have a result key greater that the previous last result key (scroll forward), or a result key smaller than the previous first result key (scroll backward). Backward scrolling has some rough edges though, so don't mind if you deliberately omitted that because the implementation is not trivial. I'm also mentioning the matter so I don't forget about it.

Reply:
Yeah I've already thought of that though I did not implement it yet.

Reply:
It's required for Jakarta Data, but not sure if I'm going to work on that immediately.